src/relations/shared_db.py

```diff
@@ -0,0 +1,182 @@
+
+    @property
+    def _peers(self):
+        return self.model.get_relation(PEER)
+
+    def _get_or_set_password(self, username: str) -> str:
+        """Retrieve password from cache or generate a new one.
+
+        Args:
+            username (str): The username.
+
+        Returns:
+            str: The password.
+        """
+        if password := self._peers.data[self._charm.app].get(f"{username}_password"):
+            return password
+        password = generate_random_password(PASSWORD_LENGTH)
+        self._peers.data[self._charm.app][f"{username}_password"] = password
+        return password
+
+    def _on_leader_elected(self, _) -> None:
+        # Ensure that the leader unit contains the latest data.
+        # Legacy apps will consume data from leader unit.
+
+        if not self._charm._is_peer_data_set:
+            # Bypass run on initial deployment leader elected event.
+            return
+
+        db_host = self._charm._mysql.get_cluster_primary_address().split(":")[0]
+
+        for relation in self.model.relations.get(LEGACY_DB_SHARED, []):
+            logger.debug(f"Syncing data from leader unit for relation {relation.id}")
+            for key, value in relation.data[self._charm.app].items():
+                if key == "db_host":
+                    relation.data[self._charm.unit][key] = db_host
+                    continue
+
+                if relation.data[self._charm.unit].get(key) != value:
+                    relation.data[self._charm.unit][key] = value
+
+    def _on_shared_db_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the legacy shared_db relation changed event.
+
+        Generate password and handle user and database creation for the related application.
+        """
+        if not self._charm.unit.is_leader():
+            return
+
+        logger.warning("DEPRECATION WARNING - `shared-db` is a legacy interface")
+
+        # get relation data
+        remote_unit_data = event.relation.data.get(event.unit)
+        if not remote_unit_data:
+            # This can happen if the relation is undone before
+            # the related app become ready
+            logger.warning("No data for remote unit. Did the relation was removed?")
+            return
+
+        local_unit_data = event.relation.data[self._charm.unit]
+        local_app_data = event.relation.data[self._charm.app]
+
+        joined_unit = event.unit.name
+
+        if joined_unit in local_unit_data.get("allowed_units", ""):
+            # Test if relation data is already set for the unit
+            # and avoid re-running it
+            logger.warning(f"Unit {joined_unit} already added to relation")
+            return
+
+        # retrieve data from the relation databag
+        # Abstracted is the fact that it is received from the leader unit of the related app
+        # fallback to peer data if data is not set for the unit (non leader unit)
```

Comment:
I was a little confused by this comment about `fallback to peer data`. Shouldn't it be `fallback to local application data`?